pkg/helper/backendapi_entity.go

```diff
@@ -0,0 +1,360 @@
+		if metric.Element.SystemName == systemName {
+			return metric.Element.ID, nil
+		}
+	}
+
+	return -1, nil
+}
+
+//
+// PRIVATE
+//
+//
+
+func (b *BackendAPIEntity) resetMethods() {
+	b.metricsAndMethods = nil
+	b.methods = nil
+}
+
+func (b *BackendAPIEntity) resetMetrics() {
+	b.metricsAndMethods = nil
+	b.metrics = nil
+}
+
+func (b *BackendAPIEntity) resetMappingRules() {
+	b.mappingRules = nil
+}
+
+func (b *BackendAPIEntity) getMethods() (*threescaleapi.MethodList, error) {
+	b.logger.V(1).Info("getMethods")
+	hitsID, err := b.getHitsID()
+	if err != nil {
+		return nil, err
+	}
+	methodList, err := b.client.ListBackendapiMethods(b.backendAPIObj.Element.ID, hitsID)
+	if err != nil {
+		return nil, fmt.Errorf("backend [%s] get methods: %w", b.backendAPIObj.Element.SystemName, err)
+	}
+	sanitizeBackendMethodList(methodList)
+
+	return methodList, nil
+}
+
+func (b *BackendAPIEntity) getMetrics() (*threescaleapi.MetricJSONList, error) {
+	b.logger.V(1).Info("getMetrics")
+	metricsAndMethods, err := b.MetricsAndMethods()
+	if err != nil {
+		return nil, err
+	}
+	metricsAndMethodsKeys := make([]string, 0, len(metricsAndMethods.Metrics))
+	for _, metric := range metricsAndMethods.Metrics {
+		metricsAndMethodsKeys = append(metricsAndMethodsKeys, metric.Element.SystemName)
+	}
+
+	methods, err := b.Methods()
+	if err != nil {
+		return nil, err
+	}
+	methodsKeys := make([]string, 0, len(methods.Methods))
+	for _, method := range methods.Methods {
+		methodsKeys = append(methodsKeys, method.Element.SystemName)
+	}
+
+	metricKeys := ArrayStringDifference(metricsAndMethodsKeys, methodsKeys)
+	metricList := &threescaleapi.MetricJSONList{
+		Metrics: make([]threescaleapi.MetricJSON, 0, len(metricKeys)),
+	}
+	for _, systemName := range metricKeys {
+		for _, metric := range metricsAndMethods.Metrics {
+			if metric.Element.SystemName == systemName {
+				metricList.Metrics = append(metricList.Metrics, metric)
+			}
+		}
+	}
+
+	return metricList, nil
+}
+
+func (b *BackendAPIEntity) getMetricsAndMethods() (*threescaleapi.MetricJSONList, error) {
+	b.logger.V(1).Info("getMetricsAndMethods")
+	metricList, err := b.client.ListBackendapiMetrics(b.backendAPIObj.Element.ID)
+	if err != nil {
+		return nil, fmt.Errorf("backend [%s] get metrics: %w", b.backendAPIObj.Element.SystemName, err)
+	}
+
+	sanitizeBackendMetricList(metricList)
+	return metricList, nil
+}
+
+func (b *BackendAPIEntity) getHitsID() (int64, error) {
+	b.logger.V(1).Info("getHitsID")
+	list, err := b.MetricsAndMethods()
+	if err != nil {
+		return 0, err
+	}
+
+	for _, metric := range list.Metrics {
+		if metric.Element.SystemName == "hits" {
+			return metric.Element.ID, nil
+		}
+	}
+
+	return 0, fmt.Errorf("backend [%s] hits not found", b.backendAPIObj.Element.SystemName)
+}
+
+// sanitizeBackendMetricList sanitizes systemName from backend metrics
+// Update is made in place
+// "system_name": "hits.45498" -> "system_name": "hits"
```

Comment:
I'm curious about this. Why is a number received from the response?

Reply:
3scale API design issues. 3scale Apisonator does not know about backends, so product metrics and backend metrics called `hits` or any other conflicting name have to be differentiated.

Reply:
:+1: